pkg/forth/forth.go

```diff
@@ -0,0 +1,274 @@
+
+// Pop pops the stack. If the stack is Empty Pop will panic.
+// Eval recovers() the panic.
+func (f *forthstack) Pop() (ret string) {
+
+	if len(f.stack) < 1 {
+		panic(errors.New("Empty stack"))
+	}
+	ret = f.stack[len(f.stack)-1]
+	f.stack = f.stack[0 : len(f.stack)-1]
+	//fmt.Printf("Pop: %v stack %v\n", ret, f.stack)
+	return ret
+}
+
+// Length returns the stack length.
+func (f *forthstack) Length() int {
+	return len(f.stack)
+}
+
+// Empty is a convenience function synonymous with Length == 0
+func (f *forthstack) Empty() bool {
+	return len(f.stack) == 0
+}
+
+// errRecover converts panics to errstr iff it is an os.Error, panics
+// otherwise.
+func errRecover(errp *error) {
+	e := recover()
+	if e != nil {
+		if _, ok := e.(runtime.Error); ok {
+			panic(e)
+		}
+		*errp = e.(error)
+	}
+}
+
+/* iEval takes a Forth and strings and splits the string on space
+ * characters, pushing each element on the stack or invoking the
+ * operator if it is found in the opmap.
+ */
+func iEval(f Forth, s string) {
+	// TODO: create a separator list based on isspace and all the
+	// non alpha numberic characters in the opmap.
+	for _, val := range strings.Fields(s) {
+		//fmt.Printf("eval %s stack %v", val, f.Stack())
+		fun := opmap[val]
+		if fun != nil {
+			//fmt.Printf("Eval ...:")
+			fun(f)
+			//fmt.Printf("Stack now %v", f.Stack())
+		} else {
+			f.Push(val)
+			//fmt.Printf("push %s, stack %v", val, f.Stack())
+		}
+	}
+	return
+}
+
+/* Eval takes a Forth and strings and splits the string on space
+ * characters, pushing each element on the stack or invoking the
+ * operator if it is found in the opmap. It returns TOS when it is done.
+ * it is an error to leave the stack non-Empty.
+ */
+func Eval(f Forth, s string) (ret string, err error) {
+	defer errRecover(&err)
+	iEval(f, s)
+	ret = f.Pop()
+	return
+
+}
+
+func toInt(f Forth) int {
+	i, err := strconv.Atoi(f.Pop())
+	if err != nil {
+		panic(err)
+	}
+	return i
+}
+
+func plus(f Forth) {
+	x := toInt(f)
+	y := toInt(f)
+	z := x + y
+	f.Push(strconv.Itoa(z))
+}
+
+func times(f Forth) {
+	x := toInt(f)
+	y := toInt(f)
+	z := x * y
+	f.Push(strconv.Itoa(z))
+}
+
+func sub(f Forth) {
+	x := toInt(f)
+	y := toInt(f)
+	z := y - x
+	f.Push(strconv.Itoa(z))
+}
+
+func div(f Forth) {
+	x := toInt(f)
+	y := toInt(f)
+	z := y / x
+	f.Push(strconv.Itoa(z))
+}
+
+func mod(f Forth) {
+	x := toInt(f)
+	y := toInt(f)
+	z := y % x
+	f.Push(strconv.Itoa(z))
+}
+
+func roundup(f Forth) {
+	rnd := toInt(f)
+	v := toInt(f)
+	v = ((v + rnd - 1) / rnd) * rnd
+	f.Push(strconv.Itoa(v))
+}
+
+func swap(f Forth) {
+	x := f.Pop()
+	y := f.Pop()
+	f.Push(x)
+	f.Push(y)
+}
+
+func strcat(f Forth) {
+	x := f.Pop()
+	y := f.Pop()
+	f.Push(y + x)
+}
+
+func dup(f Forth) {
+	x := f.Pop()
+	f.Push(x)
+	f.Push(x)
+}
+
+func ifelse(f Forth) {
+	x := toInt(f)
+	y := f.Pop()
+	z := f.Pop()
+	if x != 0 {
+		f.Push(y)
+	} else {
+		f.Push(z)
+	}
+}
+
+func hostname(f Forth) {
+	h, err := os.Hostname()
+	if err != nil {
```

Comment:
do you have to check if h is empty? or is that fine?

Reply:
if hostname returns an empty string instead of an error we are so screwed. I can do this but ... it makes no sense.